api/src/main/resources/workbench.yaml

```diff
@@ -967,6 +995,59 @@ definitions:
         format: int64
         description: Milliseconds since the UNIX epoch.
 
+  MaterializeCohortRequest:
+    type: object
+    properties:
+      cohortName:
```

Comment:
A cohort has a name and an id which aren't documented. By default, I would have expected the `cohortId` to be canonical?

Reply:
The cohort ID is a database ID, and won't be visible to researchers in our UI. The expectation is that when they reference a cohort in a notebook, it will be by name... not only because they can do that without finding a database ID, but because it makes the notebook code human-readable.

Note this means that if they change the name of a cohort, their notebook may stop working... we should have a warning about this. (We considered having a immutable-but-human-readable ID for cohorts based on the entered cohort name to avoid this, but that has problems of its own... so we decided just to have them refer to the cohort by name.)

Reply:
Is a name unique within a workspace? Also following this logic, why use workspace ID rather than workspace name in the URL? Is this consistent with Cohorts?

My usual preference would be to have a single ID in the API for referring to a resource, so I would use `id` and treat `name` as display-only in the UI or stop returning id in the API in favor of name throughout.

Reply:
Yes, name is going to be unique within a workspace. (Though I don't think we're enforcing that yet.)

The workspace ID and namespace are taken from Firecloud, and will be extracted from the REST path. The "ID" in FC is actually entered by users and is somewhat human readable (and immutable.) We didn't want to follow that pattern for AofU objects, since not being able to edit names or enter names that can't show up in REST paths is painful for users. (FC may eventually support having editable display names on workspaces, but it doesn't right now.)

In any case, the client library we provide for calling this API from a notebook can use the workspace namespace and ID for the workspace the notebook resides in, without it being present in notebook code... so researchers shouldn't have to worry about that part. They just need to build a MaterializeCohortRequest.

If you ask researchers to use IDs in their notebook code, they first have to go look up those IDs somewhere, and then remember what those IDs mean later. All in all, we thought it was better to use names.

Also: we want researchers to be able to clone workspaces (including the cohorts and notebooks within them) and have them continue to work. So if we were to use an ID here rather than a name, we'd have to make sure it was the same ID across workspaces. We don't have such an ID right now.

Reply:
This argument can be made for most APIs that have some UI component and I generally haven't agreed with this in the past.

I would argue for:
- A single consistent way of referencing a resource throughout the API (exception: searching for that resource). For example, the way you access GetCohort should be the same as how you access it here - shouldn't `GetCohort` work from a notebook just as easily? Shouldn't it work across cloned workspaces?
- For programmatic access, use IDs intended for programmatic access

I don't know that I would expect someone to remember the exact capitalization/spacing on their cohort name to begin with. In most cases they might copy-pasting anyways. Or alternatively, load all cohorts into a dataframe and let the user pull out the ID themselves. Or support search by name in the `ListCohorts` API (possibly via client library). Lastly you could change the ID to be friendlier - generate an ID based on the name and make this clear in the UI, or constrain the name by regex to something usable in a URL.

Anyways - this probably involves work beyond the scope of this PR, but would definitely definitely prefer changing this in the future.

Reply:
So the relevant notebook code options are:

1) What I'm proposing here:

api.materializeCohort("My Diabetics Cohort"...)

2) Developer looks up ID, hardcodes it in the notebook:

MY_DIABETIC_COHORT_ID = 123456789
api.materializeCohort(MY_DIABETIC_COHORT_ID....)

3) Developer programatically looks up a cohort by name, uses its ID in the notebook:

cohort = api.lookupCohort("My Diabetic Cohort")
api.materializeCohort(cohort.cohortId....)


#2 requires more work for the developer, and is a bit harder to read, but protects them against future cohort name changes.

#1 is a more convenient version of #3. Otherwise they seem equivalent. I don't see any reason to make a developer do multiple API calls to materialize a cohort by name.

In either case, there's the risk of typos when copy/pasting the cohort ID or name into the notebook code. We can minimize this risk by having a menu option for inserting cohort names or ID constants straight into the notebook code.

WRT cloning workspaces: cohorts referenced in a notebook should be cohorts defined in the same workspace; when we clone a workspace we clone the cohorts and notebooks within it, too. That way if you want to tweak the cohort subsequently (either in its definition or review) you can do that and have it take effect in the notebook; and you can make tweaks to the notebook as well. 

So if we were going to have the ID hardcoded in the notebook, we'd either need to have an ID which is scoped to the workspace (and would be the same in another workspace, which is hard to guarantee if it's just some number)  or we'd have to modify the notebook code to reference a new cohort ID in the cloned version (yuck).

We also probably want a way to import a notebook into an existing workspace, and have it reference some cohort that may not exist yet... we wouldn't know its ID, but we could know a name of a cohort you have to define to make the notebook work.

Can you elaborate on your concerns with #1? Is this just an API principle thing?

@jmandel let me know if you have thoughts on this.

Reply:
Mostly API principle, but there are implications. On principle it seems inconsistent/surprising to define a standard ID for accessing a cohort (and used for GetCohort), then instead create a secondary lookup via freeform text name on this other related method. In practice, it may be surprising (and undesirable) that if I rename something that looks like a freeform display name in the Cohort editor, that I would break a bunch of notebooks. I also find the API less consistent, which hurts comprehension for developers - but that my be subjective.

In general I'd prefer (3) of these options perhaps with some library code, if we can't just make a user defined/generated readable ID for the cohort instead. For that matter, could you just associate a cohort with a notebook at notebook creation time and wire it up through an environment variable or a default generated codeblock - as Mark was doing with the workspace? I don't remember specifics. If this is a relationship we store on our side, then we can handle clone (though explicit API calls in notebooks wouldn't carry over properly).

Reply:
When it comes to associating a cohort with a notebook: it's impossible to say (in general) whether the notebook will exist first, or the cohort. There are use cases in each direction...

I don't love the idea of using display names in the API on principole, but I think it's a pretty good tradeoff here. One other thing to consider is defining two methods in the client library, both backed by the same API:

```
api.materializeCohortById(cohort.cohortId....)

# This function does a lookup by name, and then internally calls `materializeCohortById` if there is exactly one match, or throws an exception otherwise. The REST API itself only needs to support `materializeCohortById`.
api.materializeCohort("Cohort Name")

```

Reply:
+1 to Josh's comments -- we can't expect cohorts to already exist when defining a notebook. It might be nice as a convenience to let you create a notebook with some code auto-generated based on cohorts you select in the UI, but the more flexible approach is to let you use menu options to insert code inside the notebook UI itself.

As far as making a user-readable-but-auto-generated immutable ID, that's where we started with cohorts: "My Cohort" might have an ID of "MY_COHORT". The advantage there is you don't have to worry about a rename breaking your notebook. But we ultimately decided not to do that because as soon as you allow modifications of the name or definition of a cohort, that ID may no longer be accurate. So instead we said we'd let you change the names of cohorts, with a warning that doing so might break your notebooks. If that does happen, it will be obvious enough to researchers and not that hard to fix.

Think of a cohort name like a file name included with an import statement. You want it to be human readable and understandable; you can also change the name of the file later (because sometimes that's appropriate), and that will break your code. Then you get to go fix it.

WRT to #1 vs. #3... they are functionally exactly the same. Either way, if the cohort with the name you provided doesn't exist, you'll get an error in your notebook. 

The difference is #3 is more code and more RPCs. We could certainly support passing in a cohort ID in this manner, but I'd say we'd still want to support passing in a cohort name to the materialization endpoint for convenience's sake.